src/v/archival/tests/archival_metadata_stm_gtest.cc

```diff
@@ -200,14 +204,23 @@ TEST_F_CORO(
 
     ASSERT_TRUE_CORO(!res);
 
-    auto [plagued_node, delay_applied] = co_await with_leader(
-      10s, [](raft::raft_node_instance& node) {
-          raft::response_delay fail{
-            .length = 100ms, .on_applied = ss::promise<>{}};
+    ss::shared_promise<> may_resume_append;
```

Comment:
this looks like a safe use of `shared_promise` but we usually try to avoid this type as much as possible as we've been bit by its semantics in the past.

in this case, i wonder, if you can avoid it by using a semaphore?

```c++
seastar::semaphore latch(0);

// callback
[&latch] { return latch.wait(); }

// signal
latch.signal(max_int{});
```

Reply:
It would work indeed but I feel like there is an additional mental overhead for reading the code when using a semaphore. "What does it count?", "To what value it is incremented?", etc. The shared promise approach describes it's presence in a much more straightforward way: "You can set me (only) once, and you can wait for the value to be set as many times as you want and from as many places as you need".

I probed multiple synchronization primitives and shared promise seems best suited. std::atomic_flag comes close in semantics, except the "set once" _requirement_.